File: sort/radixsort/radix_sort.cpp
#include <vector>
#include "radix_sort.h"
#include "sort/countingsort/counting_sort.hpp"

// digits means the number of digits for ints in arr.
void radix_sort(std::vector<int>& arr, int digits) {
  for (int i = 0; i < digits; i++) {
    // key_fn will extract the ith digit (0-base) of a given int.
    auto key_fn = [i](int e) {
      for (int pos = i; pos > 0; pos--) e /= 10;
      return e % 10;
    };
    std::vector<int> out;
    counting_sort(arr, out, 9, key_fn);
    arr.swap(out);
  }
}

File: sort/radixsort/py/radix_sort.py
import sort.count.py.counting_sort as csort

# digits means the number of digits for ints in A.
def sort(A, digits):
    for i in range(digits):
        A = csort.sort(A, 9, lambda a: (a / 10 ** i) % 10)
    return A
